### PR TITLE
Add TimerClient, TimerJob payload, TimerAPIError, docs, and tests

### PR DESCRIPTION
**Pull request by rudyardrichter**
* Add a client for the Timer service
    * Add `TimerClient` class, along with `TimerJob` for constructing data
      to pass to the Timer service for job creation, and ``TimerAPIError``
    * Modify `globus_sdk.config` utilities to provide URLs for Actions and
      Timer services

The docs additions might need some RST/sphinx proofreading—I'll check out the built version for the PR and see if it works.

### CODE REVIEW

**Comment by rudyardrichter** [comment]
Let me know if there's any docs additions I missed. I think I covered it, but I'm not sure if there are other places like `index.rst` where I should also link out to Timer now.

Also, I have a larger functional example script which runs most of the methods on the `TimerClient`, where would be a good place to put that?

**Comment by rudyardrichter** [comment]
(Force-pushing a bit to try to fix up the docs.)

File: src/globus_sdk/services/timer/data.py
Comment (sirosen) [inline src/globus_sdk/services/timer/data.py:80]: I think we also want to support dict data, since transfer submission supports that. It will complicate things regarding `environment` support (see the note below), but I'm looking for the SDK to generally "feel consistent" about accepting `dict | <fancy-helper>` in all locations where it's reasonable to do so. It's also an important escape-valve if the helper class, in this case `TransferData`, doesn't support some valid but unexpected format for the data.

```suggestion
        transfer_data: Union[TransferData, Dict[str, Any]],
```

File: src/globus_sdk/services/timer/data.py
Comment (sirosen) [inline src/globus_sdk/services/timer/data.py:83]: I think we should not use `**kwargs` and instead expand out the full list of args (exactly as in `__init__`).

It's somewhat annoying as a maintainer, but it has good effects on the type annotations because the signature for this method will then contain things like `name: str | None`. Without doing this, a caller can pass `name=0` here and it will pass `mypy`, but then behave in a bad way at runtime.

File: src/globus_sdk/services/timer/data.py
Comment (sirosen) [inline src/globus_sdk/services/timer/data.py:78]: I'm leaving my comments below because they'll show my thought process on this a bit, but I'd also like to offer an alternative way forward. I have questions/concerns about handling environments nicely, and I want various changes which will be tedious/laborious.

If this doesn't seem like a quick solve and we want to unblock things, we can remove this helper and add the following method:

```python
@classmethod
def get_transfer_action_url(
    cls, *, environment: Optional[str] = None
) -> str:
    return slash_join(
        get_service_url("actions", environment=environment), "transfer/transfer/run"
    )
```

plus an example added to the `docs/examples/` dir which shows how to use this to construct a timer task.

That will set us up to implement something that does this assembly step but we can think about how to handle env names being set at runtime without blocking the rest of the work.

This isn't the only way forward. If there's a way of refactoring this which satisfies my desire to support setting `environment` at runtime, I'm okay with keeping this more sophisticated helper.

File: src/globus_sdk/services/timer/errors.py
Comment (sirosen) [inline src/globus_sdk/services/timer/errors.py:15]: I don't think you need to set this classvar at all. It's not necessary to remove it, but if you want to drop a couple of lines I think it's fine to omit.

File: src/globus_sdk/services/timer/errors.py
Comment (sirosen) [inline src/globus_sdk/services/timer/errors.py:56]: Not important, but this triggered an idea which I don't want to forget. If you wanted to build a jsonpath from `loc`, you could do
```python
path = "$"
for elem in e["loc"]:
    if isinstance(elem, int):
        path += "[" + str(elem) + "]"
    else:
        path += "." + elem
```

We could add this to `utils` as a helper for all data which has path-like elements like this.

File: src/globus_sdk/_testing/data/timer/get_job.py
Comment (sirosen) [reply src/globus_sdk/_testing/data/timer/get_job.py:1]: I think it's fine to not try to share data between the fixtures, at least for now. We could maybe work something out later.

I assume the noqa is for line length?
Could we maybe dedent a bit and avoid that by assigning to some vars which we then use in the fixture? Like

```python
_ACTION_DATA = {
    "data": {
        ...
    }
}
```
?

**Review by sirosen — CHANGES_REQUESTED** [review]
With only a couple of exceptions this looks merge-ready! 👍

My major area of interest is the `TimerJob` object, and the classmethod for converting a `TransferData`. I have detailed comments there.

One meta comment/question: should there also be an `ActionsClient` object? If there's going to be a future one, that may have bearing on `TimerJob`. e.g. Perhaps `get_transfer_timer_callback_url` is a method which belongs on `ActionsClient`, not on `TimerJob`. (I haven't thought that through thoroughly.)

File: src/globus_sdk/_testing/data/timer/get_job.py
Comment (rudyardrichter) [reply src/globus_sdk/_testing/data/timer/get_job.py:1]: Yes, exactly. I'll give that a shot.

File: src/globus_sdk/_testing/models.py
Comment (rudyardrichter) [reply src/globus_sdk/_testing/models.py:16]: I was thinking about that. Maybe a stub `ActionsClient` would be reasonable. However, for the callback URL issue, it would still be a little odd, I think? Since there are several different actions, which have differing URLs. I think we'd need something like an Action Provider enum which stores a bunch of the URLs for various action providers. What do you think?

File: src/globus_sdk/config/environments.py
Comment (rudyardrichter) [reply src/globus_sdk/config/environments.py:107]: No, that seems good. Since at minimum timer + actions + flows have this in common I think it makes sense. I'll make the change.

File: src/globus_sdk/services/timer/client.py
Comment (rudyardrichter) [reply src/globus_sdk/services/timer/client.py:60]: Yeah, that would be nice. So, these will be left in for this PR, and we'll expect to remove them at some point?

File: src/globus_sdk/services/timer/data.py
Comment (rudyardrichter) [reply src/globus_sdk/services/timer/data.py:104]: What if this just includes an `environment` kwarg?

File: src/globus_sdk/services/timer/data.py
Comment (rudyardrichter) [reply src/globus_sdk/services/timer/data.py:83]: Hah, I started with kwargs as a hack, and found pylint complained about the docstring (apparently it recently changed that `\**kwargs` is the correct way to write it for sphinx?), so I was distracted by my little kwargs win. Yeah, I agree expanding the args is preferable.

File: src/globus_sdk/services/timer/errors.py
Comment (rudyardrichter) [reply src/globus_sdk/services/timer/errors.py:15]: Hmm, sure. I wrote this in partly to indicate that the attr on the base class would not be correct. This version seems slightly more correct to me in that respect (even if it isn't used)? I would set it to `None` but that would be a type error.

Comment (sirosen) [reply src/globus_sdk/services/timer/errors.py:15]: Yeah, I thought that might be the reasoning here -- it's not used but this is the only way to write "there are no `MESSAGE_FIELDS`". Makes sense. Let's mark this resolved; I'm good with keeping it.

I just wanted to make sure you knew it's not technically necessary.

File: src/globus_sdk/services/timer/data.py
Comment (sirosen) [reply src/globus_sdk/services/timer/data.py:104]: I thought about that too, as a possible solution.

I'm just not sure what the right interface is here. A `TransferData` object actually has a relevant environment because it has a `submission_id`. So, just to write down what I'm thinking, if you are setting the environment at runtime, usage might be like so...
```python
env: str = get_env_from_config_file()

# not looking things up, usages might be a little incorrect
ac = ConfidentialAppAuthClient(
    client_id=creds[env].client_id,
    client_secret=creds[env].client_secret,
    environment=env
)

tc = TransferClient(
    authorizer=ClientCredentialsAuthorizer(ac, TransferClient.scopes.all),
    environment=env
)

tdata = TransferData(tc, ...)
tdata.add_item(...)

timerc = TimerClient(
    authorizer=ClientCredentialsAuthorizer(ac, TimerClient.scopes.timer),
    environment=env
)
hourly_job = TimerJob.from_transfer_data(tdata, utcnow(), 3600, environment=env)

timerc.create_job(hourly_job)
```

Looking at this, I see three things.
1. You need to specify `environment=env` _many times_, and if you make a mistake things will not work in somewhat opaque ways
2. Each time, it's for a client init except for `TimerJob`, so that still stands out to me
3. The interface in this case is already not _great_, so maybe it doesn't matter much to make it slightly worse in the name of simplicity

In that light, this seems not that critical to me to try to get this "perfect" because it will be a little bit awkward no matter what we do. `environment: str | None = None` as a new parameter seems like the easiest solution.

**Comment by rudyardrichter** [comment]
Ok, I think I've addressed all the suggestions, and this should be good for re-review. I tested the non-prod-env case again and it's still good.

Also, I added an example page in the docs with the script I mentioned above.

File: src/globus_sdk/services/timer/data.py
Comment (sirosen) [inline src/globus_sdk/services/timer/data.py]: Let's tweak to put this line up with the `:param:`.

File: src/globus_sdk/services/timer/data.py
Comment (sirosen) [inline src/globus_sdk/services/timer/data.py]: Maybe format with
```
.. warning:
    some text
```

so that this is very attention-get-y. Otherwise, users are probably going to try `__init__` and get confused

File: src/globus_sdk/services/timer/data.py
Comment (sirosen) [inline src/globus_sdk/services/timer/data.py:111]:
```suggestion
        :param scope: Timer defaults to the Transfer 'all' scope. Use this parameter to
            change the scope used by Timer when calling the Transfer Action Provider
```

File: docs/examples/timer_operations.rst
Comment (sirosen) [inline docs/examples/timer_operations.rst:41]: After discussion, we agree that this might "explain better" if in the example,

```
interval = timedelta(minutes=30)
stop_after_n = 2
```

so that we don't need to explain the `n=1` case here.

But wait, there's more! We should also explain the `n=1` behavior in the TimerJob docs.

File: docs/examples/timer_operations.rst
Comment (sirosen) [inline docs/examples/timer_operations.rst:44]: Why don't we print the job_id here, so that delete can happen later?

File: src/globus_sdk/services/timer/data.py
Comment (sirosen) [inline src/globus_sdk/services/timer/data.py:33]: I just noticed this, but should this be documented as `timedelta or int` and "The interval in seconds ..."?
Otherwise, the only way I'd know what `interval=300` means is by reading source here and seeing that the delta is converted to seconds.

Also applies to the same param in the Transfer helper.

**Comment by rudyardrichter** [comment]
Made the updates discussed earlier today— let me know in particular if the new version of the example script makes sense.

**Comment by sirosen** [comment]
The revised example script looks great to me -- I realize the login flow makes it a bit longer, but I think it's worth it.

I've gone back over each of the things we discussed and everything seems to be resolved.
I'm going to go ahead and merge, and we can always make further adjustments if it proves necessary.

**Review by sirosen — APPROVED** [review]

**Pull request merged**
